presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/JoinPlannerUtils.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.cost.PlanNodeStatsEstimate;
+import com.facebook.presto.cost.StatsCalculator;
+import com.facebook.presto.cost.StatsProvider;
+import com.facebook.presto.sql.analyzer.FeaturesConfig.JoinDistributionType;
+import com.facebook.presto.sql.planner.iterative.Rule.Context;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+
+import static com.facebook.presto.SystemSessionProperties.getBroadcastJoinMaxTableSize;
+import static com.facebook.presto.SystemSessionProperties.getJoinDistributionType;
+import static com.facebook.presto.sql.analyzer.FeaturesConfig.JoinDistributionType.AUTOMATIC_RESTRICTED;
+
+final class JoinPlannerUtils
+{
+    private JoinPlannerUtils() {}
+
+    /**
+     * Returns true if join build side is small enough to be qualified for replication.
+     * Provided stats calculator should not underestimate statistics.
+     */
+    public static boolean canReplicate(JoinNode joinNode, StatsCalculator statsCalculator, Context context)
+    {
+        PlanNode buildSide = joinNode.getRight();
+        PlanNodeStatsEstimate buildSideStatsEstimate = calculateStatsRecursive(buildSide, statsCalculator, context);
+        double buildSideSizeInBytes = buildSideStatsEstimate.getOutputSizeInBytes(buildSide.getOutputSymbols(), context.getSymbolAllocator().getTypes());
+        JoinDistributionType joinDistributionType = getJoinDistributionType(context.getSession());
+        return joinDistributionType.canReplicate()
+                && (joinDistributionType != AUTOMATIC_RESTRICTED || buildSideSizeInBytes <= getBroadcastJoinMaxTableSize(context.getSession()).toBytes());
+    }
+
+    private static PlanNodeStatsEstimate calculateStatsRecursive(PlanNode node, StatsCalculator statsCalculator, Context context)
+    {
+        StatsProvider statsProvider = new StatsProvider()
```

Comment:
Since it is not caching, it might get expensive when applying in `ReorderJoins`. It will need to get the stats from the connector and recompute the entire tree every time `canReplicate` is called. 

 Ideally you need to add yet another stats calculator to the `Context`. But i agree that it is tedious. That's why i suggested to just add `outputOutputRowCount` to the `PlanNodeStatsEstimate`.

Reply:
Lack of caching might easily lead to O(n^2) complexity

Reply:
Added upper estimate stats provider to `Context`.